Mage.Sets/src/mage/cards/w/WeepingAngel.java

```diff
@@ -0,0 +1,156 @@
+        this.addAbility(new SpellCastOpponentTriggeredAbility(
+                new WeepingAngelMarbleizeEffect(), StaticFilters.FILTER_SPELL_A_CREATURE, false
+        ));
+
+        // If Weeping Angel would deal combat damage to a creature, prevent that damage and that creature's owner shuffles it into their library.
+        this.addAbility(new SimpleStaticAbility(new WeepingAngelDamageEffect()));
+    }
+
+    private WeepingAngel(final WeepingAngel card) {
+        super(card);
+    }
+
+    @Override
+    public WeepingAngel copy() {
+        return new WeepingAngel(this);
+    }
+}
+
+// Adapted from LoseCreatureTypeSourceEffect
+class WeepingAngelMarbleizeEffect extends ContinuousEffectImpl {
+    
+    public WeepingAngelMarbleizeEffect() {
+        super(Duration.EndOfTurn, Layer.TypeChangingEffects_4, SubLayer.NA, Outcome.Detriment);
+        staticText = "{this} isn't a creature until end of turn.";
+    }
+
+    protected WeepingAngelMarbleizeEffect(final WeepingAngelMarbleizeEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public WeepingAngelMarbleizeEffect copy() {
+        return new WeepingAngelMarbleizeEffect(this);
+    }
+
+    @Override
+    public void init(Ability source, Game game) {
+        super.init(source, game);
+        if (duration.isOnlyValidIfNoZoneChange()) {
+            // If source permanent is no longer onto battlefield discard the effect
+            if (source.getSourcePermanentIfItStillExists(game) == null) {
+                discard();
+            }
+        }
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Permanent permanent = game.getPermanent(source.getSourceId());
+        if (permanent == null) {
+            return false;
+        }
+        permanent.removeCardType(game, CardType.CREATURE);
+        if (!permanent.isTribal(game)) {
+            permanent.removeAllCreatureTypes(game);
+        }
+        if (permanent.isAttacking() || permanent.getBlocking() > 0) {
+            permanent.removeFromCombat(game);
+        }
+        return true;
+    }
+}
+
+// Based on PhyrexianVindicatorEffect
+class WeepingAngelDamageEffect extends ReplacementEffectImpl {
+
+    public WeepingAngelDamageEffect() {
+        super(Duration.WhileOnBattlefield, Outcome.PreventDamage);
+        staticText = "If {this} would deal combat damage to a creature, " +
+                "prevent that damage and that creature's owner shuffles it into their library.";
+    }
+
+    private WeepingAngelDamageEffect(final WeepingAngelDamageEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public WeepingAngelDamageEffect copy() {
+        return new WeepingAngelDamageEffect(this);
+    }
+
+    @Override
+    public boolean replaceEvent(GameEvent event, Ability source, Game game) {
+        game.preventDamage(event, source, game, Integer.MAX_VALUE);
+        Card card = game.getCard(event.getTargetId());
```

Comment:
`shuffleCardsToLibrary` takes a `Card` (or `Cards`)--is this about making sure its also a permanent before shuffling it away?

Reply:
`Permanent` extends `Card` so no issue with that method signature. It's about instanceof checks in zone handling, to perform necessary battlefield-related cleanup.

see https://github.com/magefree/mage/pull/10226#discussion_r1173222846 for example, it can cause phantom image of permanent on the battlefield